server/src/db/replay.rs

```
@@ -963,4 +1231,50 @@ mod tests {
         );
     }
+
+    #[tokio::test]
+    async fn replay_fail_lost_entry() {
+        // create write buffer state with sequence number 0 and 2, 1 is missing
+        let write_buffer_state = MockBufferSharedState::empty_with_n_sequencers(1);
+        write_buffer_state.push_entry(
+            SequencedEntry::new_from_sequence(Sequence::new(0, 0), lp_to_entry("cpu bar=1 0"))
+                .unwrap(),
+        );
+        write_buffer_state.push_entry(
+            SequencedEntry::new_from_sequence(Sequence::new(0, 2), lp_to_entry("cpu bar=1 10"))
+                .unwrap(),
+        );
+        let write_buffer = MockBufferForReading::new(write_buffer_state);
+
+        // create DB
+        let db = TestDb::builder()
+            .write_buffer(WriteBufferConfig::Reading(Arc::new(
+                tokio::sync::Mutex::new(Box::new(write_buffer) as _),
+            )))
+            .build()
+            .await
+            .db;
+
+        // construct replay plan to replay sequence numbers 0 and 1
+        let mut sequencer_numbers = BTreeMap::new();
+        sequencer_numbers.insert(0, OptionalMinMaxSequence::new(Some(0), 1));
+        let partition_checkpoint = PartitionCheckpoint::new(
+            Arc::from("table"),
+            Arc::from("partition"),
+            sequencer_numbers,
+            Utc::now(),
+        );
+        let builder = PersistCheckpointBuilder::new(partition_checkpoint);
+        let (partition_checkpoint, database_checkpoint) = builder.build();
+        let mut replay_planner = ReplayPlanner::new();
+        replay_planner.register_checkpoints(&partition_checkpoint, &database_checkpoint);
+        let replay_plan = replay_planner.build().unwrap();
+
+        // replay fails
+        let res = db.perform_replay(&replay_plan).await;
+        assert_contains!(
+            res.unwrap_err().to_string(),
+            "Cannot replay: For sequencer 0 expected to find sequence 1 but replay jumped to 2"
```

Just checking this is because the the checkpoint explicitly mentioned the seqno that doesn't exist, and not that we in general expect no gaps (as Kafka doesn't guarantee this)

---

yes, it's only when the ckpt contains a sequence that doesn't exist. The issue is the following: while replaying, we need to know when we're done. This we do by checking if the sequence number if we already reached the maximum. If we overshoot (which would happen in this case) we would either:

- consume an entry during replay that is actually out-of-scope
- seek the stream a second time
- make the stream peekable

Or instead of creating workaround, we could also fail hard, because it is very fishy when we know that there was in the past a certain sequence number that is now gone.